### PR TITLE
fix (issue#2663) : improve test coverage  

### PR DESCRIPTION
**Pull request by RishabhThakur-19**
fixed issue #2663 

# Changes

- Created `redisCompression.test.ts` to validate Redis compression and ensure lossless mapping
- Added unit tests for ArticleDetailXId
- Covered edge cases for ExtractPageTitleFromHref
- Added edge case tests for GetDefaultSkin
- Improved GetDefaultSkin to ignore unusable default skins and log appropriate warnings
- Enhanced overall test reliability and robustness

### CODE REVIEW

File: src/MediaWiki.ts [hunk 1]
```diff
@@ -409,13 +409,26 @@ class MediaWiki {
...
-    if (defaultSkins.length == 0) {
-      throw new Error(`This wiki has no default skin:\n${JSON.stringify(skins)}`)
+    // Warn if any default skins are marked unusable
+    const unusableDefaults = skins.filter((s) => s.default && s.unusable)
+    if (unusableDefaults.length > 0) {
+      logger.warn(`Ignoring unusable default skin(s): ${unusableDefaults.map((s) => s.code).join(', ')}`)
     }
+
+    // Filter usable skins first
+    const usableSkins = skins.filter((skin) => !skin.unusable)
```
Comment (Markus-Rost) [inline src/MediaWiki.ts]: `getDefaultSkin` already receives a list filtered to only usable skins.

Should a wiki actually have their default skin marked as unusable, the default value would most likely overwrite the unusable skip. However that's a misconfiguration on the wiki side anyway which we shouldn't add special support for.

Multiple default skins are technically not possible.

Comment (RishabhThakur-19) [reply src/MediaWiki.ts]: @Markus-Rost  I dont think it will bring any problem , it will only make it more robust . and 
to answer the question that should a wiki actually have their default skin as unusable ,  it is uncommon in wiki production to have a default skin as unusable , but third party or misconfigured  Mediawiki instance may mark a default as unusable  

WDYT ???

Comment (Markus-Rost) [reply src/MediaWiki.ts]: It's running the unusable filter unnecessarily for a second time and if we actually wanted to support this misconfiguration of the wiki, the correct approach would be to return the default skin even if it is marked as "unusable" instead of throwing an error like we do right now. 

Note that "unusable" just means the skin is skipped in the user preferences list, so users can't decide to use it as their personal default skin. The skin might technically still work, though custom styling might be broken as the wiki editors don't expect it being used for reads.

Comment (RishabhThakur-19) [reply src/MediaWiki.ts]: Ok I got your point, and I will bring appropriate changes to it.

File: test/unit/mwApi.test.ts [hunk 2]
```diff
@@ -210,7 +238,27 @@ describe('Mediawiki utils', () => {
         { code: 'modern', name: 'Modern', unusable: true },
       ],
     ],
-  ])('Get skin', (skins: SiteInfoSkin[]) => {
+    [
+      // All skins unusable
```
Comment (benoit74) [inline test/unit/mwApi.test.ts]: This configuration does not exits "in the wild", it would make the wiki absolutely unusable, it makes no sense to adapt our code and tests for that.

File: test/unit/mwApi.test.ts [hunk 3]
```diff
@@ -210,7 +238,27 @@ describe('Mediawiki utils', () => {
...
+    [
+      // Empty array
+      [],
+    ],
+  ])('Get skin (error cases)', (skins: SiteInfoSkin[]) => {
+    expect(() => MediaWiki.getDefaultSkin(skins)).toThrow()
+  })
+  test('no default but usable skins exist', () => {
+    const skins = [
+      { code: 'vector', name: 'Vector' },
```
Comment (benoit74) [inline test/unit/mwApi.test.ts]: This configuration does not exits "in the wild", it would make the wiki unusable, it makes no sense to adapt our code and tests for that.

File: test/unit/mwApi.test.ts [hunk 4]
```diff
@@ -197,6 +209,22 @@ describe('Mediawiki utils', () => {
...
+    [
+      // Ignore unusable default, pick valid one
+      [
+        { code: 'vector-2022', name: 'Vector (2022)', default: true, unusable: true },
+        { code: 'vector', name: 'Vector Legacy', default: true },
+      ],
+      'vector',
+    ],
+    [
+      // Default exists but is unusable
```
Comment (benoit74) [inline test/unit/mwApi.test.ts]: This configuration does not exits "in the wild", it would make the wiki absolutely unusable, it makes no sense to adapt our code and tests for that.

File: test/unit/mwApi.test.ts [hunk 5]
```diff
@@ -197,6 +209,22 @@ describe('Mediawiki utils', () => {
       ],
       'vector-2022',
     ],
+    [
+      // Ignore unusable default, pick valid one
```
Comment (benoit74) [inline test/unit/mwApi.test.ts]: This configuration does not exits "in the wild", it would make the wiki absolutely unusable, it makes no sense to adapt our code and tests for that.

**Comment by RishabhThakur-19** [comment]
> I let @Markus-Rost confirm since he has more experience than I do, but to me all the tests cases you've added around skins do not exist in the wild. A Mediawiki can have only one default skin, and it must be usable. Otherwise the wiki just doesn't work. So no-one will even try to scrape it. Displaying a message in the code when there are two defaults is already not something really required at all. We should not make that logic even more complex for no added value.

@Markus-Rost, @benoit74 Thank you for pointing out that my approach had gone beyond the intended scope. I will revert all the related changes.

@Markus-Rost also pointed out that multiple default skins are technically not possible. I had tried to justify my approach based on the existing logic that seemed to handle multiple defaults before creating this PR, which led me in that direction. I appreciate the clarification and will ensure that future work stays aligned with the project requirements.

File: test/unit/mwApi.test.ts [hunk 6]
```diff
@@ -128,6 +128,18 @@ describe('mwApi', () => {
     const absoluteIndexPathNoTitle = MediaWiki.extractPageTitleFromHref('/w/index.php?action=edit&redlink=1')
     // An index.php link missing the article title
     expect(absoluteIndexPathNoTitle).toBeNull()
+
+    const fullUrl = MediaWiki.extractPageTitleFromHref('https://en.wikipedia.org/wiki/Blue_whale')
+    // Full absolute URL
+    expect(fullUrl).toBe('Blue_whale')
+
+    const encodedUrl = MediaWiki.extractPageTitleFromHref('/wiki/Blue%20whale')
```
Comment (benoit74) [inline test/unit/mwApi.test.ts]: This is not a realistic href, white spaces do not exists AFAIK in Mediawikis, they are replaced by underscores. But other characters are indeed escaped, you can test for `/wiki/Radio_Rom%C3%A2nia_Actualit%C4%83%C8%9Bi` for instance (found on Wikipedia EN).

File: Changelog [hunk 7]
```diff
@@ -10,6 +10,7 @@ Unreleased:
 * FIX: Handle categorylockdown error with placeholder instead of crashing (@triemerge #2150)
 * FIX: Improve mwUrl and adminEmail input validation and expanding test coverage for input sanitation (@RishabhThakur-19 #2657)
 * FIX: Recursively resolve CSS @import statements (@triemerge #2649)
+* FIX: Improve test coverage for MWAPI, Redis utilities, and edge-case handling (@RishabhThakur-19 #2663)
```
Comment (benoit74) [inline Changelog]: I don't think this PR is testing MWAPI anymore, and `edge-case handling` is too precise and blury at the same, testing edge cases is de-facto the purpose of testing. If you really don't know how to describe other changes, just saying `and others` is more concise and sufficient. But I feel like you can propose a better changelog

**Comment by RishabhThakur-19** [comment]
sorry for not responding early ,
I have tried to solve the requested changes from my side , if still there is any change you want , please do let me know
